new data types for OpenMS

bernt-matthias opened the pull request:
Needed here https://github.com/galaxyproteomics/tools-galaxyp/issues/266

- [ ] analysisXML
- [ ] cachedMzML
  - consists of multple files. 
- [x] `dta`, `dta2d`, `edta` 
  - [x] `dta`: 2 column space separated float/int?  https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/a09160.html
  - [x] `edta` in the examples I found tabular data with inconsistent headers `RT	mz	Int	charge	Meta2`, `RT	MZ	INT	CHARGE	rt1	mz1	int1	charge1	Rt2	Mz2	Int2	Charge2	rT3	mZ3	inT3	chargE3`
    - https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/a09164.html#ac29c69c6f4a4441f625da4eecfe033f9
  - [x] `dta2d` 3 column float/int https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/a09156.html
- [ ] fid seems to be a binary file + some text file
  - test data would be here https://github.com/OpenMS/OpenMS/tree/develop/src/tests/class_tests/openms/data/XMassFile_test (fid + acqus)
- [x] kroenik: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/a09372.html#details
- [x] `mascotxml`
- [x] ms (Internal format of sirius https://bio.informatik.uni-jena.de/repository/dist-release-local/de/unijena/bioinf/ms/sirius/4.0.1/sirius-4.0.1-manual.pdf)  -> add sirius.ms as subclass
- [x] mrm -> subclass https://github.com/galaxyproject/galaxy/pull/8109/commits/f756b718e294768faa0d1149a28097e4416a4c1b
- [x] `mztab`
  - [x] `mzTab` (v1.x designed for proteomics) https://github.com/HUPO-PSI/mzTab/blob/master/specification_document-releases/1_0-Proteomics-Release/mzTab_format_specification.pdf
  - [x] `mzTab-M` (= v2.x mzTab) https://github.com/HUPO-PSI/mzTab/blob/master/specification_document-releases/2_0-Metabolomics-Release/mzTab_format_specification_2_0-M_release.adoc#relationship-to-mztab-1.0
- [x] novor -> just map to txt
  - only for NovorAdapetr                   File to introduce customized algorithm parameters for advanced users
  - from the novor docs (customized algorithm parameters The novorFile parameter is for the most advanced users only. Novor uses advanced machine learning algorithms to learn patterns from existing data, and utilizes those patterns in its de novo sequencing. These patterns are described by thousands of parameters and stored in a .novor file. By default, one of the built-in .novor files will be chosen by the algorithm according to fragmentation and massAnalyzer. This should be sufficient for most users. However, because of the inevitable deviation in the data caused by the instruments and labs procedures, more advanced users may want to contact rapidnovor.com to train a customized .novor file for their specific instrument and lab procedure. This usually will further improve Novor's accuracy on data produced specifically in their labs.)
- [x] omssaXML: is this just idXML?
- [ ] params
- [x] paramsXML
- [x] peff https://github.com/HUPO-PSI/PEFF used in the last FileInfo test with extension fasta 
- [x] peplist
- [x] `pqp` and `osw` https://github.com/OpenMS/OpenMS/issues/4365
  - [ ] https://github.com/galaxyproject/galaxy/pull/8458 already there, or has been developed in parallel. But no sniffer. 
- [x] psms: Percolator TSV http://www.kojak-ms.org/docs/percresults.html, maybe not worth the effort since even the percolator adapter does not output table
- [ ] psq: 
- [x] qcML: just xml with root qcML? https://github.com/OpenMS/OpenMS/blob/develop/share/OpenMS/SCHEMAS/qcML_0.0.7.xsd
- [x] spec.xml
- [x] splib
- [x] sqmass sqlite https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/a09720.html#details
  - tables: `CHROMATOGRAM  PRECURSOR     RUN           SPECTRUM    DATA          PRODUCT       RUN_EXTRA `
- [x] `trafoXML`
- [x] `xquest.xml`

Maybe useful https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3518119/

bernt-matthias committed:
datatypes: new data types for OpenMS

- added MascotXML (datatype and sniffer)
- added ConsensusXML, IdXML, FeatureXML, Ms2, Wiff
- sorted the sniffers alphabetically to simplify maintenance

needed here galaxyproteomics/tools-galaxyp#266

bernt-matthias committed:
add XQuest XML data type

bernt-matthias committed:
add pqp file format

subject to https://github.com/OpenMS/OpenMS/issues/4365

bernt-matthias committed:
added mzTab(2) and TarfoXML

bernt-matthias committed:
strip spaces from lines for detecting proteomics xml files

to allow for files like https://github.com/OpenMS/OpenMS/blob/develop/src/tests/topp/FileFilter_44_output.mzML

bernt-matthias committed:
added osw data type

bernt-matthias committed:
add sqmass

bernt-matthias committed:
add Kroenik

bernt-matthias committed:
add dta, dta2d, edta

bernt-matthias committed:
add paramXML data type

bernt-matthias committed:
add XQuest Spectra data type

And search root tag not only at the beginning of the line, i.e.
and switch pattern '^<(\w*:)?ROOT' to '<(\w*:)?ROOT' and
search instead of match ("^" in combination with match was not
necessary anyway).

bernt-matthias committed:
add qcML

bernt-matthias committed:
add psms and peplist

bernt-matthias commented:
Test fails because test data is `1 2 3` which is detected as dta2d.

Could change the test data to real text, e.g. lorem ipsum ... or maybe the dta* data types are to under specified for inclusion in Galaxy?

nsoranzo commented:
> Test fails because test data is `1 2 3` which is detected as dta2d.
> 
> Could change the test data to real text, e.g. lorem ipsum ... or maybe the dta* data types are to under specified for inclusion in Galaxy?

Maybe always require the header when sniffing dta2d? The user can always force the datatype at upload or edit metadata afterwards.

bernt-matthias commented:
@nsorazano: yep. thats probably the best. I guess the same will hold for edta.

Some weeks ago I was not able to set the ftype in the autogenerated tests used in this monster  https://github.com/galaxyproteomics/tools-galaxyp/pull/415 .. now I can .. so it should be fine.

bernt-matthias committed:
dta2d, edta: remove sniffing of to generic types of these files

bernt-matthias committed:
mrm: as subclass of Tabular

bgruening commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -227,13 +227,21 @@
     <!-- Proteomics Datatypes -->
+    <datatype extension="mrm" type="galaxy.datatypes.tabular:Tabular" subclass="true"/>
+    <datatype extension="dta" type="galaxy.datatypes.proteomics:Dta" display_in_upload="true" />
+    <datatype extension="dta" type="galaxy.datatypes.proteomics:Dta" display_in_upload="true" />
```
this one is included twice

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -53,6 +53,147 @@ def generate_primary_file(self, dataset=None):
+    def sniff_prefix(self, file_prefix):
+        fh = file_prefix.string_io()
+        line = [_.strip() for _ in fh.readline().split("\t")]
+        if len(line) != len(self.column_names) or line != self.column_names:
```
line != self.column_names should be enough isn't it?

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -53,6 +53,147 @@ def generate_primary_file(self, dataset=None):
+    def sniff_prefix(self, file_prefix):
+        fh = file_prefix.string_io()
+        line = [_.strip() for _ in fh.readline().split("\t")]
+        if len(line) == 5 and line == self.column_names:
```
line == self.column_names is enough here?

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -53,6 +53,147 @@ def generate_primary_file(self, dataset=None):
+    def sniff_prefix(self, file_prefix):
+        fh = file_prefix.string_io()
+        line = [_.strip() for _ in fh.readline().split("\t")]
+        if len(line) == 6 and line == self.column_names:
```
line == self.column_names enough?

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -92,6 +233,277 @@ def display_peek(self, dataset):
+@build_sniff_from_prefix
+class Dta(TabularData):
+    """dta
+    The first line contains the singly protonated peptide mass (MH+) and the
```
single?

bernt-matthias replied:
Don't know. Copy pasted it from here: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/classOpenMS_1_1DTAFile.html. Could also replace the text with the link.

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -92,6 +233,277 @@ def display_peek(self, dataset):
+        data_row = []
+        data_lines = 0
+        if dataset.has_data():
+            with open(dataset.file_name, 'r') as dtafile:
```
how much data is in such a file, can it be GB?

bernt-matthias replied:
I have no Idea -- but its a good idea to make this assumption. 

Maybe https://github.com/galaxyproject/galaxy/blob/80607164f4b33d4df0f0a7c2cc8f2e65a1b8e8ec/lib/galaxy/datatypes/data.py#L843 would be a solution. Since the `set_meta` function is of `TabularData` not implemented I can't use it via inheritance. 

Maybe it would be nice to implement this function as decorator or a utility function?

Note that csv / tsv types have the same problem https://github.com/galaxyproject/galaxy/blob/80607164f4b33d4df0f0a7c2cc8f2e65a1b8e8ec/lib/galaxy/datatypes/tabular.py#L940

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -92,6 +233,277 @@ def display_peek(self, dataset):
+        dataset.metadata.delimiter = " "
+
+    def sniff_prefix(self, file_prefix):
+        has_data = False
```
I would remove this sniffer here, this seems to be too loosely defined.

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -92,6 +233,277 @@ def display_peek(self, dataset):
+
+    def _parse_header(self, line):
+        if len(line) != 3 or len(line[0]) < 3 or not line[0].startswith("#"):
+            return None
```
Why not return False here?

bernt-matthias replied:
Could change this, but I found it more appealing to return a list (if successful) or None (otherwise). I also check for `None` in the remaining code.

bgruening commented on lib/galaxy/datatypes/proteomics.py:
```diff
@@ -218,6 +650,29 @@ class UniProtXML(ProteomicsXml):
+class QCML(ProteomicsXml):
+    """qcml
+    https://github.com/OpenMS/OpenMS/blob/develop/share/OpenMS/SCHEMAS/mzQCML_0_0_5.xsd
+    https://github.com/OpenMS/OpenMS/blob/develop/share/OpenMS/SCHEMAS/qcML_0.0.7.xsd
```
should this be a link to a specific commit ...?

bernt-matthias replied:
OK. Will do - even if there is versioning in the filename :)

bgruening reviewed (commented):
@bernt-matthias thanks a lot! I have added a few comments inline.

It would also be good to include some doc-tests I think, in the spirit of https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/binary.py#L1634

bernt-matthias committed:
suggested changes from review

- no sniffing for dta
- remove superficial len checks
- schema links to OpenMS with commit

bernt-matthias committed:
sniffer tests for openms related data types

bernt-matthias committed:
openms datatypes: fix mzTab

bernt-matthias committed:
MzTab: fix capitalization

bernt-matthias committed:
remove subclass relation between OSW and PQP

bernt-matthias committed:
PQP: just check that the extra OSW tables are not included

bernt-matthias committed:
disable sublass="true" for osw, pqp, sqmass

bernt-matthias committed:
remove sqlite data types with sniffers from registry

bernt-matthias committed:
move SQlite datatype up in the datatypes_config.sample

mvdbeek commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -265,6 +265,7 @@
     <datatype extension="mzdata" type="galaxy.datatypes.proteomics:MzData" mimetype="application/xml" display_in_upload="true"/>
     <datatype extension="ms2" type="galaxy.datatypes.proteomics:Ms2" display_in_upload="true"/>
     <datatype extension="mzq" type="galaxy.datatypes.proteomics:MzQuantML" mimetype="application/xml" display_in_upload="true"/>
+    <datatype extension="sqlite" type="galaxy.datatypes.binary:SQlite" mimetype="application/octet-stream" display_in_upload="true"/>
```
You have to move the sniffer up as well

mvdbeek replied:
Nevermind, of course that'd not right

bernt-matthias committed:
remove db3 from registry

bernt-matthias committed:
match thermo.raw extension as in datatypes_conf

bernt-matthias committed:
add sirius.ms datatype

as subclass of text

bernt-matthias committed:
remove duplicate osw, pqp entries in conf

bernt-matthias committed:
add PEFF data type

bernt-matthias commented:
Hi @bgruening except for datatypes derived from ProteomicsXml all new data types have now working working doc-tests. 

I'm hesitant (aka. lazy) to add doc-tests for the ProteomicsXml cases .. what do you think?

bgruening commented:
That is super fine by me.
Lets ask for a review from @nsoranzo or @mvdbeek. This is really great work! Thanks a lot!
